webapi/src/main/scala/org/knora/webapi/responders/v2/ontology/DeleteCardinalitiesFromClass.scala

```diff
@@ -0,0 +1,396 @@
+
+      currentTime: Instant = Instant.now
+
+      updateSparql = org.knora.webapi.messages.twirl.queries.sparql.v2.txt
+                       .replaceClassCardinalities(
+                         triplestore = settings.triplestoreType,
+                         ontologyNamedGraphIri = internalOntologyIri,
+                         ontologyIri = internalOntologyIri,
+                         classIri = internalClassIri,
+                         newCardinalities = newInternalClassDefWithLinkValueProps.directCardinalities,
+                         lastModificationDate = deleteCardinalitiesFromClassRequest.lastModificationDate,
+                         currentTime = currentTime
+                       )
+                       .toString()
+
+      _ <- (storeManager ? SparqlUpdateRequest(updateSparql)).mapTo[SparqlUpdateResponse]
+
+      // Check that the ontology's last modification date was updated.
+
+      _ <- OntologyHelpers.checkOntologyLastModificationDateAfterUpdate(
+             settings,
+             storeManager,
+             internalOntologyIri = internalOntologyIri,
+             expectedLastModificationDate = currentTime,
+             featureFactoryConfig = deleteCardinalitiesFromClassRequest.featureFactoryConfig
+           )
+
+      // Check that the data that was saved corresponds to the data that was submitted.
+
+      loadedClassDef <- OntologyHelpers.loadClassDefinition(
+                          settings,
+                          storeManager,
+                          classIri = internalClassIri,
+                          featureFactoryConfig = deleteCardinalitiesFromClassRequest.featureFactoryConfig
+                        )
+
+      _ = if (loadedClassDef != newInternalClassDefWithLinkValueProps) {
+            throw InconsistentRepositoryDataException(
+              s"Attempted to save class definition $newInternalClassDefWithLinkValueProps, but $loadedClassDef was saved"
+            )
+          }
+
+      // Update the cache.
+
+      updatedOntology = ontology.copy(
+                          ontologyMetadata = ontology.ontologyMetadata.copy(
+                            lastModificationDate = Some(currentTime)
+                          ),
+                          classes = ontology.classes + (internalClassIri -> readClassInfo)
+                        )
+
+      _ = Cache.storeCacheData(
+            cacheData.copy(
+              ontologies = cacheData.ontologies + (internalOntologyIri -> updatedOntology)
+            )
+          )
+
+      // Read the data back from the cache.
+
+      response: ReadOntologyV2 <- OntologyHelpers.getClassDefinitionsFromOntologyV2(
+                                    classIris = Set(internalClassIri),
+                                    allLanguages = true,
+                                    requestingUser = deleteCardinalitiesFromClassRequest.requestingUser
+                                  )
+
```

Comment:
You have forgotten a very important step here: to update the cardinalities of subclasses.  If you are removing cardinalities from base class, the subclasses should be updated.

Reply:
@SepidehAlassi done, I think. Could you please check if this is what you meant?